maze_dataset/tokenization/util.py

```diff
@@ -146,3 +151,125 @@ def coords_to_strings(
+                adj_list[i, 1] = c_s
+
+            i += 1
+
+    if shuffle_d0:
+        np.random.shuffle(adj_list)
+
+    return adj_list
+
+
+def equal_except_adj_list_sequence(rollout1: list[str], rollout2: list[str]) -> bool:
+    """Returns if the rollout strings are equal, allowing for differently sequenced adjacency lists.
+    <ADJLIST_START> and <ADJLIST_END> tokens must be in the rollouts.
+    Intended ONLY for determining if two tokenization schemes are the same for rollouts generated from the same maze.
+    This function should NOT be used to determine if two rollouts encode the same `LatticeMaze` object.
+
+    # Warning: CTT False Positives
+    This function is not robustly correct for some corner cases using `CoordTokenizers.CTT`.
+    If rollouts are passed for identical tokenizers processing two slightly different mazes, a false positive is possible.
+    More specifically, some cases of zero-sum adding and removing of connections in a maze within square regions along the diagonal will produce a false positive.
+    """
+
+    def get_token_regions(toks: list[str]) -> tuple[list[str], list[str]]:
+        adj_list_start, adj_list_end = toks.index("<ADJLIST_START>") + 1, toks.index(
+            "<ADJLIST_END>"
+        )
+        adj_list = toks[adj_list_start:adj_list_end]
+        non_adj_list = toks[:adj_list_start] + toks[adj_list_end:]
+        return adj_list, non_adj_list
+
+    if len(rollout1) != len(rollout2):
+        return False
+    if ("<ADJLIST_START>" in rollout1) ^ ("<ADJLIST_START>" in rollout2):
+        return False
+    if ("<ADJLIST_END>" in rollout1) ^ ("<ADJLIST_END>" in rollout2):
+        return False
+
+    adj_list1, non_adj_list1 = get_token_regions(rollout1)
+    adj_list2, non_adj_list2 = get_token_regions(rollout2)
+    if non_adj_list1 != non_adj_list2:
+        return False
+    counter1: Counter = Counter(adj_list1)
+    counter2: Counter = Counter(adj_list2)
+    return counter1 == counter2
+
+
+def flatten(it: Iterable[any], levels_to_flatten: int | None = None) -> Generator:
+    """
+    Flattens an arbitrarily nested iterable.
+    Flattens all iterable data types except for `str` and `bytes`.
+
+    # Returns
+    Generator over the flattened sequence.
+
+    # Parameters
+    - `it`: Any arbitrarily nested iterable.
+    - `levels_to_flatten`: Number of levels to flatten by. If `None`, performs full flattening.
+    """
+    for x in it:
+        # TODO: swap type check with more general check for __iter__() or __next__() or whatever
+        if (
+            hasattr(x, "__iter__")
+            and not isinstance(x, (str, bytes))
+            and (levels_to_flatten is None or levels_to_flatten > 0)
+        ):
+            yield from flatten(
+                x, None if levels_to_flatten == None else levels_to_flatten - 1
+            )
+        else:
+            yield x
+
+
+def get_all_subclasses(class_: type, include_self=False) -> set[type]:
```

Comment:
minor type hint nitpick: this should use `typing.Type` and look like
`def get_all_subclasses(class_: Type, include_self: bool = False) -> set[Type]:`